File: umc1-14/src/main/java/cam72cam/mod/entity/EntityRegistry.java
package cam72cam.mod.entity;

import cam72cam.mod.ModCore;
import cam72cam.mod.event.ClientEvents;
import cam72cam.mod.event.CommonEvents;
import cam72cam.mod.resource.Identifier;
import cam72cam.mod.text.PlayerMessage;
import cam72cam.mod.world.World;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screen.DisconnectedScreen;
import net.minecraft.client.gui.screen.MainMenuScreen;
import net.minecraft.client.gui.screen.MultiplayerScreen;
import net.minecraft.entity.EntityClassification;
import net.minecraft.entity.EntityType;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.registries.ForgeRegistries;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class EntityRegistry {
    private static final Map<Class<? extends Entity>, EntityType<?>> registered = new HashMap<>();

    private static String missingResources;

    private EntityRegistry() {

    }

    public static void register(ModCore.Mod mod, Supplier<Entity> ctr, EntitySettings settings, int distance) {
        Entity tmp = ctr.get();
        Class<? extends Entity> type = tmp.getClass();
        Identifier id = new Identifier(mod.modID(), type.getSimpleName());

        // TODO expose updateFreq and vecUpdates

        CommonEvents.Entity.REGISTER.subscribe(() -> {
            EntityType.Builder<net.minecraft.entity.Entity> builder = EntityType.Builder.create((et, world) -> new ModdedEntity(et, world, ctr, settings), EntityClassification.MISC)
                    .setShouldReceiveVelocityUpdates(false)
                    .setTrackingRange(distance)
                    .setUpdateInterval(20)
                    .setCustomClientFactory((se, world) -> new ModdedEntity(registered.get(type), world, ctr, settings));
            if (settings.immuneToFire) {
                builder = builder.immuneToFire();
            }
            EntityType<net.minecraft.entity.Entity> et = builder.build(id.toString());
            et.setRegistryName(id.internal);
            ForgeRegistries.ENTITIES.register(et);
            registered.put(type, et);
        });
    }

    public static Entity create(World world, Class<? extends Entity> cls) {
        return ((ModdedEntity)registered.get(cls).create(world.internal)).getSelf();
    }

    public static void registerEvents() {
        CommonEvents.Entity.REGISTER.subscribe(() -> ForgeRegistries.ENTITIES.register(SeatEntity.TYPE));
        CommonEvents.Entity.JOIN.subscribe((world, entity) -> {
            if (entity instanceof ModdedEntity) {
                if (World.get(world) != null) {
                    String msg = ((ModdedEntity) entity).getSelf().tryJoinWorld();
                    if (msg != null) {
                        //missingResources = msg;
                        //return false; GAH FORGE FMLPlayMessages:166!
                    }
                }
            }
            return true;
        });
    }

    @OnlyIn(Dist.CLIENT)
    public static void registerClientEvents() {
        ClientEvents.TICK.subscribe(() -> {
            if (missingResources != null && !Minecraft.getInstance().isSingleplayer() && Minecraft.getInstance().getConnection() != null) {
                ModCore.error(missingResources);
                Minecraft.getInstance().getConnection().getNetworkManager().closeChannel(PlayerMessage.direct(missingResources).internal);
                Minecraft.getInstance().loadWorld(null);
                Minecraft.getInstance().displayGuiScreen(new DisconnectedScreen(new MultiplayerScreen(new MainMenuScreen()), "disconnect.lost", PlayerMessage.direct(missingResources).internal));
                missingResources = null;
            }
        });
    }
}

File: src/main/kotlin/org/eln2/mc/common/entities/EntityRegistry.kt
package org.eln2.mc.common.entities

import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.MobCategory
import net.minecraft.world.level.Level
import net.minecraftforge.eventbus.api.IEventBus
import net.minecraftforge.registries.DeferredRegister
import net.minecraftforge.registries.ForgeRegistries
import net.minecraftforge.registries.RegistryObject
import org.eln2.mc.MODID

object EntityRegistry {
    val ENTITY_REGISTRY = DeferredRegister.create(ForgeRegistries.ENTITY_TYPES, MODID)!!

    fun setup(bus: IEventBus) {
        ENTITY_REGISTRY.register(bus)
    }

    inline fun <reified T : Entity> entity(
        id: String,
        category: MobCategory,
        crossinline supplier: (b: EntityType<T>, l: Level) -> T,
        crossinline configure: (EntityType.Builder<T>) -> Unit,
    ): RegistryObject<EntityType<T>> = ENTITY_REGISTRY.register(id) {
        EntityType.Builder.of({ t, l -> supplier(t, l) }, category).also(configure).build(id)
    }

    inline fun <reified T : Entity> entity(
        id: String,
        category: MobCategory,
        crossinline supplier: (b: EntityType<T>, l: Level) -> T,
    ) = entity(id, category, supplier) { }
}
